packages/protocol/contracts-0.8/common/UsingPrecompiles.sol

```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
    * @return Epoch number.
    */
   function getEpochNumberOfBlock(uint256 blockNumber) public view returns (uint256) {
-    return epochNumberOfBlock(blockNumber, getEpochSize());
+    if (isL2()) {
+      return getEpochManager().getEpochNumberOfBlock(blockNumber);
+    } else {
+      return epochNumberOfBlock(blockNumber, getEpochSize());
+    }
   }
 
   /**
    * @notice Returns the epoch number at a block.
    * @return Current epoch number.
    */
   function getEpochNumber() public view returns (uint256) {
-    return getEpochNumberOfBlock(block.number);
+    if (isL2()) {
+      return getEpochManager().getCurrentEpochNumber();
+    } else {
+      return getEpochNumberOfBlock(block.number);
+    }
   }
 
   /**
-   * @notice Gets a validator address from the current validator set.
+   * @notice Gets a validator signer address from the current validator set.
    * @param index Index of requested validator in the validator set.
-   * @return Address of validator at the requested index.
+   * @return Address of validator signer at the requested index.
    */
   function validatorSignerAddressFromCurrentSet(
     uint256 index
   ) public view virtual returns (address) {
-    bytes memory out;
-    bool success;
-    (success, out) = GET_VALIDATOR.staticcall(abi.encodePacked(index, uint256(block.number)));
-    require(success, "error calling validatorSignerAddressFromCurrentSet precompile");
-    return address(uint160(getUint256FromBytes(out, 0)));
+    if (isL2()) {
+      return getAccounts().getValidatorSigner(validatorAddressFromCurrentSet(index));
+    } else {
+      bytes memory out;
+      bool success;
+      (success, out) = GET_VALIDATOR.staticcall(abi.encodePacked(index, uint256(block.number)));
+      require(success, "error calling validatorSignerAddressFromCurrentSet precompile");
+      return address(uint160(getUint256FromBytes(out, 0)));
+    }
   }
 
   /**
-   * @notice Gets a validator address from the validator set at the given block number.
+   * @notice Gets a validator address from the current validator set.
    * @param index Index of requested validator in the validator set.
-   * @param blockNumber Block number to retrieve the validator set from.
    * @return Address of validator at the requested index.
    */
+  function validatorAddressFromCurrentSet(uint256 index) public view onlyL2 returns (address) {
+    address[] memory _elected = getEpochManager().getElected();
```

Comment:
More gas optimization: add an indexed function to EpochManager that reads only the specific index, rather than reading the whole array into memory.